lib/spack/spack/util/web.py

```diff
@@ -336,161 +309,194 @@ def list_url(url):
+        set of visited links.
+    """
+    # Cache of visited links, meant to be captured by the closure below
+    _visited = set()
+
+    def _spider(url, root, collect_nested):
+        """Fetches URL and any pages it links to.
+
+        Prints out a warning only if the root can't be fetched; it ignores
+        errors with pages that the root links to.
+
+        Args:
+            url (str): url being fetched and searched for links
+            root (str): base url on which we want to restrict our search
+            collect_nested (bool): whether we want to collect arguments
+                for nested spidering on the links found in this url
+
+        Returns:
+            A tuple of:
+            - pages: dict of pages visited (URL) mapped to their full text.
+            - links: set of links encountered while visiting the pages.
+            - spider_args: argument for subsequent call to spider
+        """
+        pages = {}  # dict from page URL -> text content.
+        links = set()  # set of all links seen on visited pages.
+        subcalls = []
 
-        if subcalls:
-            pool = NonDaemonPool(processes=len(subcalls))
-            try:
-                results = pool.map(_spider_wrapper, subcalls)
+        try:
+            response_url, _, response = read_from_url(url, 'text/html')
+            if not response_url or not response:
+                return pages, links, subcalls
 
-                for sub_pages, sub_links in results:
-                    pages.update(sub_pages)
-                    links.update(sub_links)
+            page = codecs.getreader('utf-8')(response).read()
+            pages[response_url] = page
 
-            finally:
-                pool.terminate()
-                pool.join()
+            # Parse out the links in the page
+            link_parser = LinkParser()
+            link_parser.feed(page)
 
-    except URLError as e:
-        tty.debug(e)
+            while link_parser.links:
+                raw_link = link_parser.links.pop()
+                abs_link = url_util.join(
+                    response_url,
+                    raw_link.strip(),
+                    resolve_href=True)
+                links.add(abs_link)
 
-        if hasattr(e, 'reason') and isinstance(e.reason, ssl.SSLError):
-            tty.warn("Spack was unable to fetch url list due to a certificate "
-                     "verification problem. You can try running spack -k, "
-                     "which will not check SSL certificates. Use this at your "
-                     "own risk.")
+                # Skip stuff that looks like an archive
+                if any(raw_link.endswith(s) for s in ALLOWED_ARCHIVE_TYPES):
+                    continue
 
-        if raise_on_error:
-            raise NoNetworkConnectionError(str(e), url)
+                # Skip things outside the root directory
+                if not abs_link.startswith(root):
+                    continue
```

Comment:
@tgamblin @adamjstewart When debugging I noticed that here we are testing if `abs_link` (a string) starts with a `ParseResult` object, which will always be false. Do we want to fix the logic, or delete the test (and thus keep the current actual behavior)?

Reply:
If it's always false, I would delete the check to speed things up. But I would need to take a deeper look at the code to make sure it isn't needed.

Reply:
The intent here was to avoid leaving the root subdirectories (e.g. going to entirely different websites), so fixing the check would likely speed things up by not exploring links that go off the root site. 

It might be worth running `spack versions` for each package and seeing how much this helps and whether it affects the output.  I can't decide what to do because there might be packages where spelunking off the root site is needed.  I kind of doubt it, but fixing this would make it so we couldn't get to those pages.  Would it affect github releases?  Do those links go outside the root page?  If so then we should probably delete the check.

Reply:
I just checked: fixing the check makes 15 tests fail and several packages are finding less versions (e.g. `boost`). I'll delete the check for now, which means I'll keep the current behavior. Let me know if you want me to open an issue to revisit this logic later and optimize the search.

Reply:
See e85b6fe